PocketKit/Tests/PocketKitTests/PocketLoggedOutViewModelTests.swift

```diff
@@ -5,40 +5,46 @@ import AuthenticationServices
 
 
 class PocketLoggedOutViewModelTests: XCTestCase {
-    private var viewModel: PocketLoggedOutViewModel!
-    private var subscriptions: Set<AnyCancellable>!
+    private var authorizationClient: AuthorizationClient!
     private var appSession: AppSession!
+    private var subscriptions: Set<AnyCancellable>!
 
     @Published
     var foo: String = ""
 
     override func setUp() {
+        authorizationClient = AuthorizationClient(
+            consumerKey: "test-consumer-key",
+            authenticationSession: MockAuthenticationSession.self
+        )
         appSession = AppSession(keychain: MockKeychain())
         subscriptions = []
     }
 
     override func tearDown() {
         subscriptions = []
     }
+
+    func subject(
+        authorizationClient: AuthorizationClient? = nil,
+        appSession: AppSession? = nil
+    ) -> PocketLoggedOutViewModel {
+        let viewModel = PocketLoggedOutViewModel(
+            authorizationClient: authorizationClient ?? self.authorizationClient,
+            appSession: appSession ?? self.appSession
+        )
+        viewModel.contextProvider = self
+        return viewModel
+    }
 }
 
 extension PocketLoggedOutViewModelTests {
     func test_logIn_onFxAError_setsPresentedAlert() {
-        let urlSession = MockURLSession()
-        urlSession.stubData { request in
-            self.validGUIDResponse(for: request)
-        }
-
-        let client = AuthorizationClient(
+        let failingClient = AuthorizationClient(
```

Comment:
Unfortunate that we need to re-create the `AuthorizationClient` here so that we can inject a different type of mock. It would be nice if we just had one `MockAuthenticationSession` that we can manipulate instead of having to rebuild the whole dependency graph.

I see why we need this: we only pass in the _type_ of auth session, not a concrete instance. The factory pattern can help us here. If we inject a function that creates the `AuthenticationSession` instead just a simple type.

I'd like to see this get done before merging the PR, but I could be convinced to leave it as-is for now if it's going to be a time sink.

Reply:
@jacobsimeon I'd like to discuss this one live, if possible.